### ISSUE TITLE
Sg3Reader 

### ISSUE THREAD

**Issue by PeterC28** [bvschaik/julius#623]
I am trying to create a project which opens and displays the graphics for Pharaoh, and found the "SG File Format" and SG Reader code on GitHub and have been using them as a guide. 

1. I have found a few differences between the 2 formats, mainly in the Image Info records. Which is the more up to date format?
2. I don't see anywhere in the code a place which uses the image Info data and creates the actual image.s Am I missing something?

Thanks.

**Comment by bvschaik** [comment]
1. SG3 is the "later" format, but which one you should use depends on the game. Caesar 3 ships with .sg**2** files and uses SG2; Pharaoh has .sg**3** files so it's using SG3.
2. Take a look at [SgImage::getImage()](https://github.com/bvschaik/citybuilding-tools/blob/master/sgreader/sgimage.cpp#L116): that's reading the file, creating a QImage and setting pixels on it in `set555Pixel()`

By the way, do let me know if anything is unclear in the SG File Format page, so I can make it better.

**Comment by PeterC28** [comment]
1. So, which is the SG2 and which is the SG3 format? and the differences depend on the version number, correct?

**Comment by bvschaik** [comment]
The differences depend on the version number, indeed. The two "alpha only" fields apply to v214 (Zeus and Emperor), they're not present in C3 (sg2) or Pharaoh (sg3). Actually, I'm not sure what the differences between v211 (sg2) and v212 (sg3) are.

**Comment by PeterC28** [comment]
What about the ImageType field? In the code it's a UInt16 and the documentation has it as a byte followed by a one byte "Flag: image data is fully compressed". Is it a byte for C2 (sg2) and Pharaoh (sg3) and a UInt16 for v214 (Zeus and Emperor)?

Then there's the code part:

stream->skipRawData(26);

 In the documentation you have:

3 unknown shorts
Number of animation sprites
Unknown short
X offset of animation sprite
Y offset of animation sprite
Unknown bytes
Flag: animation is reversible
Unknown byte

And lastly there's:

stream->skipRawData(7);

Where in the documentation you have:

Unknown byte
Animation speed ID
Unknown byte
Zero bytes

I'd like to apologize if I'm being a nuisance.

**Comment by bvschaik** [comment]
One important thing to keep in mind: SG Reader was written a long time ago (2007), when I was only focused on getting the images out of the .sg2/.sg3 files. While working on Julius, I got a more thorough understanding of how the image format works in Caesar 3 specifically and what some fields are used for. The documentation page was written after this time, so it includes the extra info, but I never got around to updating the SG Reader because those fields are not important for reading the images anyway.

Due to that history: I first assumed ImageType to be a uint16, later on I discovered those can be interpreted as two separate bytes with their own meaning.

Good point about indicating the signed-ness of the variables, I'll have a look at updating that.

And no problem about being a nuisance, every suggestion to improve documentation is welcome.

**Comment by PeterC28** [comment]
One mistake you made with the Data Types. The line should read:

16 | 4 | int32 | Used in SG3 only: offset into .555 of the mirrored image

The value can (and possibly must always) be negative. For example in Pharaoh's Egyptian.sg3 file the first few sprites of the warship face northeast, east, and southeast. The fifth through seventh are their mirror images (southwest, west and northwest), so the offset must be negative (-4 in this case) to get there.

**Comment by bvschaik** [comment]
Good point. I did most of the fields by memory. For the 32-bits numbers I guess it makes sense to interpret them all as `signed`, since their value can be so large.

As for your last question: last week I added a bit on the [image data](https://github.com/bvschaik/citybuilding-tools/wiki/SG-file-format#image-data), which also mentions "external" .555 files (i.e., .555 files without .sg).

**Comment by PeterC28** [comment]
So the Images associated with that bitmap would apply to the external 555 file?

Also, do you have any other game file formats and are you interested in acquiring  others?

**Comment by bvschaik** [comment]
> So the Images associated with that bitmap would apply to the external 555 file?

Yes, exactly.

> Also, do you have any other game file formats and are you interested in acquiring others?

I'm only interested in the game file formats that are used in the old citybuilding games by Impressions/Sierra.

**Comment by PeterC28** [comment]
So, Caesar 1 - 3, Pharaoh & Cleopatra, Zeus & Poseidon, Rise of the Middle Kingdom (Imp/Sierra), Outpost (Sierra)? I'll let you know if I find anything.

Do you have other formats from these games?
